contracts/smt/smt.sol

```diff
@@ -36,25 +55,36 @@ contract SMT {
         uint256 fnc;
     }
 
-    constructor(address _poseidonUnit2ContractAddr, address _poseidonUnit3ContractAddr) {
+    function initialize(
+        address _poseidonUnit2ContractAddr,
+        address _poseidonUnit3ContractAddr
+    ) public initializer {
         _poseidonUnit2 = PoseidonUnit2(_poseidonUnit2ContractAddr);
         _poseidonUnit3 = PoseidonUnit3(_poseidonUnit3ContractAddr);
+        __Ownable_init();
+    }
+
+    function getMaxDepth() public pure returns (uint256) {
+        return MAX_DEPTH;
+    }
+
+    function getRootHistory() public view returns (RootHistoryInfo[] memory) {
+        return rootHistory;
     }
 
     function add(uint256 _i, uint256 _v) public {
-        Node memory node = Node(
-            NodeType.LEAF,
-            0,
-            0,
-            _i,
-            _v
+        rootHistory.push(
+            RootHistoryInfo(root, uint64(block.timestamp), uint64(block.number))
         );
-
+        Node memory node = Node(NodeType.LEAF, 0, 0, _i, _v);
         root = addLeaf(node, root, 0);
```

Comment:
Looks like a bug. The current root is not going to be in the history. So you need to keep in mind that the last element of the array is not the current root. 
Moreover, this will put block time and number for the next root in the sequence, not the time/number which corresponds to the root hash
```
        rootHistory.push(
            RootHistoryInfo(root, uint64(block.timestamp), uint64(block.number))
        )
```

Reply:
Agree, will fix